src/ngraph/op/batch_norm.cpp

```diff
@@ -69,6 +69,69 @@ ngraph::op::BatchNorm::BatchNorm(double eps,
     add_output(input->get_element_type(), m_bn_variance_shape);
 }
 
+ngraph::op::BatchNorm::BatchNorm(bool training,
+                                 double eps,
+                                 std::shared_ptr<ngraph::Node> gamma,
+                                 std::shared_ptr<ngraph::Node> beta,
+                                 std::shared_ptr<ngraph::Node> input,
+                                 std::shared_ptr<ngraph::Node> mean,
+                                 std::shared_ptr<ngraph::Node> variance)
+    : RequiresTensorViewArgs("BatchNorm", {gamma, beta, input, mean, variance})
+    , m_bn_input_shape(input->get_shape())
+    , m_bn_variance_shape(variance->get_shape())
+    , m_bn_mean_shape(mean->get_shape())
+    , m_epsilon(eps)
+    , m_training(training)
+{
+    if (m_bn_input_shape.size() < 2)
+    {
+        throw ngraph_error("input tensor to batchnorm must have tensor of at least rank 2");
+    }
+
+    if (m_bn_input_shape[1] == 0)
+    {
+        throw ngraph_error(
+            "input tensor must have at least one channel axis for batch normalization");
+    }
+
+    auto et = input->get_element_type();
+    const char* input_names[] = {"gamma", "beta", "input", "mean", "variance"};
+
+    for (size_t i = 0; i < get_input_size(); i++)
+    {
+        if (get_input_op(i)->get_element_type() != et)
+        {
+            auto err_msg = std::string("The element type of ") + input_names[i] +
+                           " isn't equal to input data's type";
+            throw ngraph_error(err_msg.c_str());
+        }
+    }
+
+    for (size_t index = 0; index < get_input_size(); index++)
+    {
+        if (index != 2 && get_input_op(index)->get_shape().size() != 1)
```

Comment:
Would this method be a little easier to read through if you used symbolic names for the inputs?  For example:
``` c++
enum { kInputGamma = 0, kInputBeta, kInputData, kInputMean, kInputVariance };
...
if (index != kInputData && ...) {
...
```

Reply:
Heck, I'd be a fan of having these in the class's public interface.  Something like:
``` c++
class BatchnormTrainingOp ... {
   public:
      enum { kInputGamma = 0, kInputBeta, kInputData, kInputMean, kInputVariance };
      enum { kOutputData = 0, kBatchMean, kBatchVariance };
      enum { kNumInputs = kInputVariance, kNumOutputs = kBatchVariance };
   ...
};
```

Reply:
I second that. Though, I'm not sure we need a full-blown enum. A simple constant w/ a name might suffice.
```
const size_t INPUT_INDEX = 2;
```

Reply:
agreed, will make necessary changes